devel/build

```diff
@@ -47,18 +47,6 @@ docker buildx build \
     --cache-from $BASE_IMAGE \
     --cache-to type=inline \
     --tag $BASE_BUILDER_IMAGE:$tag \
-    --load \
+    --output type=oci,dest=test.tar \
```

Comment:
This is mostly an arbitrary value that supports multi-platform builds, retrieved from [the `--output` docs](https://docs.docker.com/engine/reference/commandline/buildx_build/#output). I'll refine it once the build is successful, along with trying [`docker manifest push`](https://docs.docker.com/engine/reference/commandline/manifest/#manifest-push).

Reply:
Looking into this since build is successful now. However, I don't think it's as simple as getting the push to work. Based on some reading [[1]] [[2]], it doesn't seem possible to use the built multi-arch images locally, which is what we currently do for testing before pushing.

2 options I can think of:

1. Use one `docker buildx build` command per platform. This would allow us to continue our current pattern of build > test > tag > push. However, this would go against the purpose of using `buildx` in the first place, and I think we'd need different names for each platform which is undesirable.
2. Use `--push` here to push intermediate tags, which we then pull for testing (on each platform?) before pushing with the final `$TAG` (i.e. build > tag (intermediate) > push > pull > test > tag > push). This might be best solved by splitting CI into separate GitHub Actions workflows.

I think option 2 is more appropriate.

[1]: https://github.com/docker/buildx/issues/166#issuecomment-544827163
[2]: https://developer.arm.com/documentation/102475/0100/Multi-architecture-images

Reply:
We could run a [local registry](https://hub.docker.com/_/registry) as a [sidecar service](https://docs.github.com/en/actions/using-containerized-services/about-service-containers) in CI and `--push` / `--output type=image,push=true` to that. Then we can do whatever testing we want and finally push the whole thing up to the public registry for the release phase of CI.

Reply:
@tsibley thanks, I hadn't thought of that nor used local registries before. I wonder what's the benefit of that vs. (2) above – is it just the lack of intermediate tags pushed to Docker Hub?

My main concern with (2) that also applies to your option is the expansion of `devel/build` to push to a registry, which historically has been the duty of `devel/push`. This would lead to usability of those scripts being tied to logic in the CI workflow (e.g. we'd have to set up a local registry for local development too?).

Reply:
Lack of public version of those tags, yeah. This is more important than it looks at first blush, I think, because I think all of those tags will need to be unique to a CI run to avoid collisions, so it won't just be N more mutable tags for N platforms (e.g. not just `nextstrain/base:arm64`) but an ongoing series of them (N tags × each CI run, e.g. `nextstrain/base:arm64-${uuid}`).

Agreed it sort of muddies the current clarity of separated concerns with the build steps, though I'd still see a separation where `devel/build` produces the (multi-arch) image by pushing to the local registry and `devel/push` is responsible for pushing it publicly. And yes, local dev would have to have a local registry running too (though we could make this ~easy if desired).

An alternative to the local registry would be for `devel/build` to export with `type=oci` (or similar) as you have already and then _also_ be responsible for loading that into the local Docker daemon so it's immediately usable after `devel/build`. Note that the image loaded into the local Docker daemon in this case **wouldn't** be suitable for pushing publicly, as it wouldn't be a multi-arch manifest. We'd still have to have a separate step that used `docker manifest create` on the OCI tarballs to push to the public registry.

Reply:
The advantage I see of the local registry over the alternative above is that the registry more closely matches (and thus tests) what an actual interaction looks like with the public registry and the built image. e.g. if we produce a bad multi-arch image, we're unlikely to catch it unless we're interacting with a registry in CI.

Reply:
> I think all of those tags will need to be unique to a CI run to avoid collisions

Why wouldn't this apply to the current implementation? The tag is the same for all runs on a PR, so it represents the latest run with a successful push:

https://github.com/nextstrain/docker-base/blob/c3857f55612e4ccc5834348ea51a9c57acb7f9f7/.github/workflows/ci.yml#L23

So I'd expect something like `TAG=branch-${GITHUB_REF_NAME//[^A-Za-z0-9._-]/-}-$ARCH` to behave similarly.

Regardless, I'm now experimenting with using a local registry as you described.

Reply:
These changes make use of a local registry: ecac84094b3effcac91bd098f69456f79a66fd32...9ca46fb98f441d645e99004d71521b4e7929ddde

Reply:
> Why wouldn't this apply to the current implementation? The tag is the same for all runs on a PR, so it represents the latest run with a successful push…

Hmm, yes, but I _think_ it's subtly different because instead of all CI runs for a branch sharing a single tag, they're now sharing a _set of tags_. A single tag is simple "last one to push wins" behaviour, but a set of tags will at times be a mix of pushes from different CI runs. This is maybe acceptable if pushes are the final step and all subsequent pulls are by humans who can make judgement calls (e.g. the CI run I cared about pushed and I see there aren't any other CI runs). But with each CI run also _pulling_ what it previously pushed in order to make the multi-arch image, the potential for a run to get an image from another concurrent CI run seems like a recipe for unhappiness.

Reply:
I see, that makes sense. Though that doesn't apply currently since CI is only pulling `latest`, it would apply after #57 which is something I plan to look into soon.

Reply:
Yes, but I think it would apply to using `docker manifest create` to assemble the per-arch images into a multi-arch image outside of `docker buildx build`. AIUI, `docker manifest create` would be querying the registry for the info of the per-arch images, even if it's not downloading their layers.

Reply:
I think this is resolved. Local registry is working pretty well in the latest iteration (fcc2b72f18b5ddb243dfff57c27dbc87fbfc88be...ff2edf1190bb2ce9a0323e84992a4111bb0963d7).

We don't have to mess with `docker manifest` in this approach.

Reply:
Awesome!